python/pyspark/mlv2/base.py

```diff
@@ -0,0 +1,426 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from abc import ABCMeta, abstractmethod
+
+import copy
+import threading
+
+from typing import (
+    Any,
+    Callable,
+    Generic,
+    Iterator,
+    List,
+    Optional,
+    Sequence,
+    Tuple,
+    TypeVar,
+    Union,
+    cast,
+    overload,
+    TYPE_CHECKING,
+)
+
+from pyspark import since
+from pyspark.ml.param import P
+from pyspark.ml.common import inherit_doc
+from pyspark.ml.param.shared import (
+    HasInputCol,
+    HasOutputCol,
+    HasLabelCol,
+    HasFeaturesCol,
+    HasPredictionCol,
+)
+from pyspark.sql.dataframe import DataFrame
```

Comment:
will we have a `GenericDataFrame`? @HyukjinKwon 

I feel here should be something like:
```
if not is_remote():
    from pyspark.sql.dataframe import DataFrame
else:
    from pyspark.sql.connect.dataframe import DataFrame
```

Reply:
Yeah .. I am thinking that we might need a parent class for both too .. 
BTW, cannot call `is_remote()` when importing because `SPARK_REMOTE` is set when Spark session is created.

If this is just being used for type hinting, I think just using `from pyspark.sql.dataframe import DataFrame` would be fine for now because they have the same API anyway.

Reply:
> I am thinking that we might need a parent class for both too
+1, FYI the new ML implementations will only use pure Dataframe/Column/Functions APIs, in order to support both regular spark and spark connect.

Reply:
Not a big issue, the DataFrame class is only used for type hint in the code